core/lib/zksync_core/src/vm_runner/storage.rs

```diff
@@ -0,0 +1,400 @@
+        l1_batch_number: L1BatchNumber,
+    ) -> anyhow::Result<Option<PgOrRocksdbStorage<'_>>> {
+        self.access_storage_inner(stop_receiver, l1_batch_number)
+            .await
+    }
+}
+
+/// A runnable task that catches up the provided RocksDB cache instance to the latest processed
+/// batch and then continuously makes sure that this invariant is held for the foreseeable future.
+/// In the meanwhile, `StorageSyncTask` also loads the next `max_batches_to_load` batches in memory
+/// so that they are immediately accessible by [`VmRunnerStorage`].
+#[derive(Debug)]
+pub struct StorageSyncTask<L: VmRunnerStorageLoader> {
+    pool: ConnectionPool<Core>,
+    l1_batch_params_provider: L1BatchParamsProvider,
+    chain_id: L2ChainId,
+    rocksdb_cell: Arc<OnceCell<RocksDB<StateKeeperColumnFamily>>>,
+    loader: L,
+    state: Arc<RwLock<State>>,
+    catchup_task: AsyncCatchupTask,
+}
+
+impl<L: VmRunnerStorageLoader> StorageSyncTask<L> {
+    async fn new(
+        pool: ConnectionPool<Core>,
+        chain_id: L2ChainId,
+        rocksdb_path: String,
+        loader: L,
+        state: Arc<RwLock<State>>,
+    ) -> anyhow::Result<Self> {
+        let mut conn = pool.connection_tagged(L::name()).await?;
+        let l1_batch_params_provider = L1BatchParamsProvider::new(&mut conn)
+            .await
+            .context("Failed initializing L1 batch params provider")?;
+        let rocksdb_cell = Arc::new(OnceCell::new());
+        let catchup_task = AsyncCatchupTask::new(
+            pool.clone(),
+            rocksdb_path,
+            rocksdb_cell.clone(),
+            Some(loader.latest_processed_batch(&mut conn).await?),
+        );
+        drop(conn);
+        Ok(Self {
+            pool,
+            l1_batch_params_provider,
+            chain_id,
+            rocksdb_cell,
+            loader,
+            state,
+            catchup_task,
+        })
+    }
+
+    pub async fn run(self, stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        self.catchup_task.run(stop_receiver.clone()).await?;
+        let rocksdb = self.rocksdb_cell.get().ok_or_else(|| {
+            anyhow::anyhow!("Expected RocksDB to be initialized by `AsyncCatchupTask`")
+        })?;
+        loop {
+            if *stop_receiver.borrow() {
+                tracing::info!("`StorageSyncTask` was interrupted");
+                return Ok(());
+            }
+            // It's important to hold a lock on `state` while we are updating RocksDB cache
+            // as otherwise `VmRunnerStorage` will have an inconsistent view on DB state.
+            let mut state = self.state.write().await;
+            let mut conn = self.pool.connection_tagged(L::name()).await?;
+            let latest_processed_batch = self.loader.latest_processed_batch(&mut conn).await?;
+            let rocksdb_builder = RocksdbStorageBuilder::from_rocksdb(rocksdb.clone());
+            if rocksdb_builder.l1_batch_number().await == Some(latest_processed_batch) {
```

Comment:
l1_batch_number() returns last batch persisted in rocksdb + 1, you should take in into account

Reply:
Thanks, I keep forgetting about this quirk!

I will try to add more tests to cover various corner cases.